### PR TITLE
Auto trash emptying

### LINKED ISSUE #986: Auto empty trash for "My Library"
Auto empty trashed items after 30 days

### PR DESCRIPTION
**Pull request by michalrentka**
Closes #986

### CODE REVIEW

File: Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift [hunk 1]
```diff
@@ -16,27 +16,49 @@ struct GeneralSettingsView: View {
...
+                Text(L10n.Settings.General.afterXDays(7)).tag(7)
+                Text(L10n.Settings.General.afterXDays(15)).tag(15)
+                Text(L10n.Settings.General.afterXDays(30)).tag(30)
+                Text(L10n.Settings.General.never).tag(0)
+            }
         }
         .navigationBarTitle(L10n.Settings.General.title)
     }
+
+    private func thresholdString(for threshold: Int) -> String {
```
Comment (mvasilak) [inline Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift:45]: This seems not used anywhere.

File: Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift [hunk 2]
```diff
@@ -16,27 +16,49 @@ struct GeneralSettingsView: View {
...
+            return L10n.Settings.General.afterXDays(threshold)
+        }
+    }
 }
 
 struct GeneralSettingsView_Previews: PreviewProvider {
     static var previews: some View {
         GeneralSettingsView()
     }
 }
```
Comment (mvasilak) [inline Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift:63]: The preview crashes, due to the lack of environment object instances. If we don't care for it, we can probably ommit it completely.

File: Zotero/Assets/en.lproj/Localizable.strings [hunk 3]
```diff
@@ -291,6 +291,10 @@
 "settings.general.show_subcollections_title" = "Show Items from Subcollections";
 "settings.general.show_collection_item_counts" = "Show collection sizes";
 "settings.general.open_links_in_external_browser" = "Open links in external browser";
+"settings.general.autoempty_title" = "Delete Items in Trash";
+"settings.general.after_one_day" = "After 1 Day";
+"settings.general.never" = "Never";
+"settings.general.after_x_days" = "After %d Days";
```
Comment (mvasilak) [inline Zotero/Assets/en.lproj/Localizable.strings:297]: `settings.general.after_one_day` and `settings.general.after_x_days` can be implemented as following in `Localizable.stringsdict` instead:
```
<key>settings.general.after_x_days</key>
    <dict>
        <key>NSStringLocalizedFormatKey</key>
        <string>%#@after_x_days@</string>
        <key>after_x_days</key>
        <dict>
            <key>NSStringFormatSpecTypeKey</key>
            <string>NSStringPluralRuleType</string>
            <key>NSStringFormatValueTypeKey</key>
            <string>d</string>
            <key>one</key>
            <string>After 1 Day</string>
            <key>other</key>
            <string>After %d Days</string>
        </dict>
    </dict>
```

and use `L10n.Settings.General.afterXDays(1)`

File: Zotero/Models/Defaults.swift [hunk 4]
```diff
@@ -69,6 +69,12 @@ final class Defaults {
     @UserDefault(key: "QuickCopyAsHtml", defaultValue: false, defaults: .standard)
     var quickCopyAsHtml: Bool
 
+    @UserDefault(key: "TrashAutoEmptyDayThreshold", defaultValue: 30, defaults: .standard)
+    var trashAutoEmptyThreshold: Int
+
+    @UserDefault(key: "TrashLastAutoEmptyDate", defaultValue: .distantPast, defaults: .standard)
+    var lastAutoEmptyDate: Date
```
Comment (mvasilak) [inline Zotero/Models/Defaults.swift:76]: Why not keep the pattern?

```suggestion
    var trashLastAutoEmptyDate: Date
```

**Review by mvasilak — CHANGES_REQUESTED** [review]
Since the defaults are initialized with 30 days and distant past, first time the app is launched with this, it will empty trash. I guess that is desired, and will also set a proper date for last trash emptying.

File: Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift (hunk 1 above)
Comment (michalrentka) [reply Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift]: Removed 5ae4b8c5aff31c30fcbbffc5e535f1eb194f18f7

File: Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift (hunk 2 above)
Comment (michalrentka) [reply Zotero/Scenes/Master/Settings/General/Views/GeneralSettingsView.swift]: Removed 5fa7b54d04667ecdd7cfbf67b31319499833d003

File: Zotero/Models/Defaults.swift (hunk 4 above)
Comment (michalrentka) [reply Zotero/Models/Defaults.swift]: Changed fab280cb166d1f3787bd44eeab8f136df4332457

File: Zotero/Assets/en.lproj/Localizable.strings (hunk 3 above)
Comment (michalrentka) [reply Zotero/Assets/en.lproj/Localizable.strings]: Yeah I forgot we have this implementation already, fixed in c13bb883f24e478da1b01f80a83eb975854b0c96

**Review by mvasilak — APPROVED** [review]

**Review by mvasilak — APPROVED** [review]

**Pull request merged**